apps/fz_wall/lib/fz_wall/cli/helpers/nft.ex

```diff
@@ -0,0 +1,161 @@
+  require Logger
+  @table_name "firezone"
+
+  @doc """
+  Creates a nft rule
+  """
+  def create_rule(type, source_set, dest_set, action) do
+    exec!("""
+      #{nft()} 'add rule inet #{@table_name} forward #{rule_match_str(type, source_set, dest_set, action)}'
+    """)
+  end
+
+  @doc """
+  Removes a nft rule
+  """
+  def remove_rule(type, source_set, dest_set, action) do
+    delete_rule_matching(rule_match_str(type, source_set, dest_set, action))
+  end
+
+  @doc """
+  Adds an element from a nft set
+  """
+  def add_elem(set, ip) do
+    exec!("""
+      #{nft()} 'add element inet #{@table_name} #{set} { #{standardized_inet(ip)} }'
+    """)
+  end
+
+  @doc """
+  Deletes an element from a nft set
+  """
+  def delete_elem(set, ip) do
+    exec!("""
+      #{nft()} 'delete element inet #{@table_name} #{set} { #{standardized_inet(ip)} }'
+    """)
+  end
+
+  @doc """
+  Adds a nft set
+  """
+  def add_set(set_spec) do
+    exec!("""
+      #{nft()} 'add set inet #{@table_name} #{set_spec.name} { type #{set_type(set_spec.type)} ; flags interval ; }'
+    """)
+  end
+
+  @doc """
+  Deletes a nft set
+  """
+  def delete_set(set_spec) do
+    exec!("""
+      #{nft()} 'delete set inet #{@table_name} #{set_spec.name}'
+    """)
+  end
+
+  @doc """
+  Sets up firezone table.
+  """
+  def setup_table do
+    exec!("#{nft()} create table inet #{@table_name}")
+  end
+
+  @doc """
+  Sets up firezone chains.
+  """
+  def setup_chains do
+    exec!(
+      "#{nft()} 'add chain inet #{@table_name} forward " <>
+        "{ type filter hook forward priority 0 ; policy accept ; }'"
+    )
+
+    exec!(
+      "#{nft()} 'add chain inet #{@table_name} postrouting " <>
+        "{ type nat hook postrouting priority 100 ; }'"
+    )
+
+    # XXX: Do more testing with this method of creating masquerade rules
+    for int <- File.ls!("/sys/class/net/") do
+      # Masquerade all interfaces except loopback and our own wireguard interface
+      if int not in ["lo", wireguard_interface_name()] do
+        exec!(
+          "#{nft()} 'add rule inet #{@table_name} postrouting oifname " <>
+            "#{int} masquerade persistent'"
+        )
+      end
+    end
+  end
+
+  @doc """
+  Deletes nft tables (This will remove)
+  """
+  def teardown_table do
+    if table_exists?() do
+      exec!("#{nft()} delete table inet #{@table_name}")
+    end
+  end
+
+  defp nft do
+    Application.fetch_env!(:fz_wall, :nft_path)
+  end
+
+  defp table_exists? do
+    cmd = "#{nft()} list table inet #{@table_name}"
+
+    case bash(cmd) do
+      {_result, 0} ->
+        true
+
+      {error, _exit_code} ->
+        if String.contains?(error, "Error: No such file or directory") do
+          false
+        else
+          raise """
+            Unknown Error from command #{cmd}. Error:
+            #{error}
+          """
+        end
+    end
+  end
+
+  defp delete_rule_matching(rule_str) do
+    rules = exec!("#{nft()} -a list table inet #{@table_name}")
```

Comment:
Hmm... I think we're calling this each time for every `device, rule` combination for a user when a user is deleted. I wonder if we improve this by moving the `list table` exec outside this function.

Reply:
Same reasoning as [here](https://github.com/firezone/firezone/pull/766#discussion_r910415170) the handle might change.

 This is only called 6 times each time a user is deleted, 1 for each nftable rule, meaning each combination of ip type and action for a user, no matter how many devices/rules a user have.